controller/composite/controller.go

```diff
@@ -67,13 +71,14 @@ type parentController struct {
 	updateStrategy updateStrategyMap
 	childInformers common.InformerMap
 
-	numWorkers int
+	numWorkers    int
+	eventRecorder record.EventRecorder
 
 	finalizer *finalizer.Manager
 	customize customize.Manager
 }
 
-func newParentController(resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcClient mcclientset.Interface, revisionLister mclisters.ControllerRevisionLister, cc *v1alpha1.CompositeController, numWorkers int) (pc *parentController, newErr error) {
+func newParentController(resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcClient mcclientset.Interface, revisionLister mclisters.ControllerRevisionLister, cc *v1alpha1.CompositeController, numWorkers int, eventRecorder record.EventRecorder) (pc *parentController, newErr error) {
```

Comment:
Do you think we can refactor this as following:

```go
// ParentControllerConfig is used to create new instances of
// ParentController
type ParentControllerConfig {
  Field1 int
  Field2 string
  Field3 string
}

type ParentController {
  field1 int
  field2 string
  field3 string
  field4 string
}

func NewParentController(config ParentControllerConfig) (*ParentController, error) {
  // some logic ...

  pc := &ParentController{
    field1: config.Field1,
    field2: config.Field2,
    field3: config.Field3,
  }
  
  // some other logic ...
  return pc, nil
}
```

Reply:
Above pattern ensures a fixed signature. 
This reduces the impact of changes at caller code.

Reply:
I did that for `Config`,  but looking at 
```go
func newParentController(resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcClient mcclientset.Interface, revisionLister mclisters.ControllerRevisionLister, cc *v1alpha1.CompositeController, numWorkers int, eventRecorder record.EventRecorder) (pc *parentController, newErr error) {
```
I do not see clear separation what we could consider `configuration` here, except `numWorkers int` - others are objects required for running.

I believe we can think of a bit of refactoring, as most of this it taken from `mc` - 
```go
pc, err := newParentController(mc.resources, mc.dynClient, mc.dynInformers, mc.mcClient, mc.revisionLister, cc, mc.numWorkers, mc.eventRecorder)
```


but would like to do it in separate pull request, to do not mix too many changes together

Reply:
If not config then options is good naming.

Reply:
sure, however will do separate Pull Request, to not adding more changes here.

Reply:
follow up PR - in progress- https://github.com/metacontroller/metacontroller/pull/223